pkg/util/kms/kms_kmip_storage.go

```diff
@@ -0,0 +1,479 @@
+		return nil, fmt.Errorf("Unexpected operation, real %v expected %v", bi.Operation, operation)
+	}
+	if !bytes.Equal(uniqueBatchItemID, bi.UniqueBatchItemID) {
+		return nil, fmt.Errorf("Unexpected uniqueBatchItemID, real %v expected %v", bi.UniqueBatchItemID, uniqueBatchItemID)
+	}
+	if kmip14.ResultStatusSuccess != bi.ResultStatus {
+		return nil, fmt.Errorf("Unexpected result status %v expected success %v", bi.ResultStatus, kmip14.ResultStatusSuccess)
+	}
+
+	return &bi, nil
+}
+
+// KMIP handshake
+func (k *KMIPSecretStorage) discover(conn *tls.Conn) error {
+	respMsg, decoder, uniqueBatchItemID, err := k.send(conn, kmip14.OperationDiscoverVersions, kmip.DiscoverVersionsRequestPayload{
+		ProtocolVersion: []kmip.ProtocolVersion{
+			{ProtocolVersionMajor: protocolMajor, ProtocolVersionMinor: protocolMinor},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	bi, err := k.response(respMsg, kmip14.OperationDiscoverVersions, uniqueBatchItemID)
+	if err != nil {
+		return err
+	}
+
+	var respDiscoverVersionsPayload kmip.DiscoverVersionsResponsePayload
+	err = decoder.DecodeValue(&respDiscoverVersionsPayload, bi.ResponsePayload.(ttlv.TTLV))
+	if err != nil {
+		return err
+	}
+
+	if len(respDiscoverVersionsPayload.ProtocolVersion) != 1 {
+		return fmt.Errorf("Invalid len of discovered protocol versions %v expected 1", len(respDiscoverVersionsPayload.ProtocolVersion))
+	}
+	pv := respDiscoverVersionsPayload.ProtocolVersion[0]
+	if pv.ProtocolVersionMajor != protocolMajor || pv.ProtocolVersionMinor != protocolMinor {
+		return fmt.Errorf("Invalid discovered protocol version %v.%v expected %v.%v", pv.ProtocolVersionMajor, pv.ProtocolVersionMinor, protocolMajor, protocolMinor)
+	}
+	return nil
+}
+
+// GetSecret returns the secret data associated with the
+// supplied secretId.
+func (k *KMIPSecretStorage) GetSecret(
+	secretID string,
+	keyContext map[string]string,
+) (map[string]interface{}, error) {
+
+	log := util.Logger()
+
+	// KMIP key uniqueIdentifier
+	uniqueIdentifier, exists := k.secret.StringData[KMIPUniqueID]
+	if !exists {
+		log.Errorf("KMIPSecretStorage.GetSecret() uniqueIdentifier %v does not exist in secret %v", KMIPUniqueID, k.secret)
+		return nil, secrets.ErrInvalidSecretId
+	}
+
+	conn, err := k.connect()
+	if err != nil {
+		log.Errorf("KMIPSecretStorage.GetSecret() failed to connect %v", err)
+		return nil, err
+	}
+	defer conn.Close()
+
+	respMsg, decoder, uniqueBatchItemID, err := k.send(conn, kmip14.OperationGet, kmip.GetRequestPayload{
+		UniqueIdentifier: uniqueIdentifier,
+	})
+	if err != nil {
+		log.Errorf("KMIPSecretStorage.GetSecret() failed to send %v", err)
+		return nil, err
+	}
+	bi, err := k.response(respMsg, kmip14.OperationGet, uniqueBatchItemID)
+	if err != nil {
+		log.Errorf("KMIPSecretStorage.GetSecret() failed to verify response %v", err)
+		return nil, err
+	}
+	var getRespPayload kmip.GetResponsePayload
+	err = decoder.DecodeValue(&getRespPayload, bi.ResponsePayload.(ttlv.TTLV))
```

Comment:
```suggestion
	ttlvPayload, ok := bi.ResponsePayload.(ttlv.TTLV)
	if !ok {
		return "", errors.New("failed to parse responsePayload")
	}

	err = decoder.DecodeValue(&respRespPayload, ttlvPayload)
```

add a typecheck here ?
 
(a linter was complaining over at cephcsi)